File: rust/src/api/metrics.rs
// this file is @generated
#[allow(unused_imports)]
use crate::{error::Result, models::*, Configuration};

#[derive(Default)]
pub struct MetricsListMetricsOptions {
    pub product_family_id: Option<ProductFamilyId>,

    /// Search by metric name or code
    pub search: Option<String>,

    /// Sort order. Format: `column.direction`. Allowed columns: `name`, `code`, `created_at`. Direction: `asc` or `desc`. Default: `name.asc`.
    pub order_by: Option<String>,

    /// Page number (0-indexed)
    pub page: Option<i32>,

    /// Number of items per page
    pub per_page: Option<i32>,
}

pub struct Metrics<'a> {
    cfg: &'a Configuration,
}

impl<'a> Metrics<'a> {
    pub(super) fn new(cfg: &'a Configuration) -> Self {
        Self { cfg }
    }

    pub async fn list_metrics(
        &self,
        options: Option<MetricsListMetricsOptions>,
    ) -> Result<crate::models::MetricListResponse> {
        let MetricsListMetricsOptions {
            product_family_id,
            search,
            order_by,
            page,
            per_page,
        } = options.unwrap_or_default();

        crate::request::Request::new(http1::Method::GET, "/api/v1/metrics")
            .with_optional_query_param("product_family_id", product_family_id)
            .with_optional_query_param("search", search)
            .with_optional_query_param("order_by", order_by)
            .with_optional_query_param("page", page)
            .with_optional_query_param("per_page", per_page)
            .execute(self.cfg)
            .await
    }

    pub async fn create_metric(
        &self,
        create_metric_request: crate::models::CreateMetricRequest,
    ) -> Result<crate::models::Metric> {
        crate::request::Request::new(http1::Method::POST, "/api/v1/metrics")
            .with_body_param(create_metric_request)
            .execute(self.cfg)
            .await
    }

    pub async fn get_metric(&self, metric_id: String) -> Result<crate::models::Metric> {
        crate::request::Request::new(http1::Method::GET, "/api/v1/metrics/{metric_id}")
            .with_path_param("metric_id", metric_id)
            .execute(self.cfg)
            .await
    }

    /// Partially update metric fields. Code and aggregation_type are immutable.
    pub async fn update_metric(
        &self,
        metric_id: String,
        update_metric_request: crate::models::UpdateMetricRequest,
    ) -> Result<crate::models::Metric> {
        crate::request::Request::new(http1::Method::PATCH, "/api/v1/metrics/{metric_id}")
            .with_path_param("metric_id", metric_id)
            .with_body_param(update_metric_request)
            .execute(self.cfg)
            .await
    }

    pub async fn archive_metric(&self, metric_id: String) -> Result<()> {
        crate::request::Request::new(http1::Method::POST, "/api/v1/metrics/{metric_id}/archive")
            .with_path_param("metric_id", metric_id)
            .returns_nothing()
            .execute(self.cfg)
            .await
    }

    pub async fn unarchive_metric(&self, metric_id: String) -> Result<()> {
        crate::request::Request::new(http1::Method::POST, "/api/v1/metrics/{metric_id}/unarchive")
            .with_path_param("metric_id", metric_id)
            .returns_nothing()
            .execute(self.cfg)
            .await
    }
}

File: typescript/src/api/metrics.ts
// this file is @generated
import {
  type CreateMetricRequest,
  CreateMetricRequestSerializer,
} from "../models/createMetricRequest";
import { type Metric, MetricSerializer } from "../models/metric";
import {
  type MetricListResponse,
  MetricListResponseSerializer,
} from "../models/metricListResponse";
import type { ProductFamilyId } from "../models/productFamilyId";
import {
  type UpdateMetricRequest,
  UpdateMetricRequestSerializer,
} from "../models/updateMetricRequest";
import { HttpMethod, MeteroidRequest, type MeteroidRequestContext } from "../request";

export interface MetricsListMetricsOptions {
  productFamilyId?: ProductFamilyId;
  /** Search by metric name or code */
  search?: string;
  /** Sort order. Format: `column.direction`. Allowed columns: `name`, `code`, `created_at`. Direction: `asc` or `desc`. Default: `name.asc`. */
  orderBy?: string;
  /** Page number (0-indexed) */
  page?: number;
  /** Number of items per page */
  perPage?: number;
}

export class Metrics {
  public constructor(private readonly requestCtx: MeteroidRequestContext) {}

  /**  */
  public listMetrics(options?: MetricsListMetricsOptions): Promise<MetricListResponse> {
    const request = new MeteroidRequest(HttpMethod.GET, "/api/v1/metrics");

    request.setQueryParam("product_family_id", options?.productFamilyId);
    request.setQueryParam("search", options?.search);
    request.setQueryParam("order_by", options?.orderBy);
    request.setQueryParam("page", options?.page);
    request.setQueryParam("per_page", options?.perPage);
    return request.send(this.requestCtx, MetricListResponseSerializer._fromJsonObject);
  }

  /**  */
  public createMetric(createMetricRequest: CreateMetricRequest): Promise<Metric> {
    const request = new MeteroidRequest(HttpMethod.POST, "/api/v1/metrics");

    request.setBody(CreateMetricRequestSerializer._toJsonObject(createMetricRequest));
    return request.send(this.requestCtx, MetricSerializer._fromJsonObject);
  }

  /**  */
  public getMetric(metricId: string): Promise<Metric> {
    const request = new MeteroidRequest(HttpMethod.GET, "/api/v1/metrics/{metric_id}");

    request.setPathParam("metric_id", metricId);
    return request.send(this.requestCtx, MetricSerializer._fromJsonObject);
  }

  /** Partially update metric fields. Code and aggregation_type are immutable. */
  public updateMetric(
    metricId: string,
    updateMetricRequest: UpdateMetricRequest
  ): Promise<Metric> {
    const request = new MeteroidRequest(HttpMethod.PATCH, "/api/v1/metrics/{metric_id}");

    request.setPathParam("metric_id", metricId);
    request.setBody(UpdateMetricRequestSerializer._toJsonObject(updateMetricRequest));
    return request.send(this.requestCtx, MetricSerializer._fromJsonObject);
  }

  /**  */
  public archiveMetric(metricId: string): Promise<void> {
    const request = new MeteroidRequest(
      HttpMethod.POST,
      "/api/v1/metrics/{metric_id}/archive"
    );

    request.setPathParam("metric_id", metricId);
    return request.sendNoResponseBody(this.requestCtx);
  }

  /**  */
  public unarchiveMetric(metricId: string): Promise<void> {
    const request = new MeteroidRequest(
      HttpMethod.POST,
      "/api/v1/metrics/{metric_id}/unarchive"
    );

    request.setPathParam("metric_id", metricId);
    return request.sendNoResponseBody(this.requestCtx);
  }
}
